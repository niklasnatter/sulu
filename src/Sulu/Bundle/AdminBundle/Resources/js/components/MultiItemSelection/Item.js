// @flow
import React from 'react';
import type {Node} from 'react';
import {SortableHandle} from 'react-sortable-hoc';
import classNames from 'classnames';
import Icon from '../Icon';
import itemStyles from './item.scss';

const DRAG_ICON = 'su-more';

type Props<T> = {
    children: Node,
    disabled: boolean,
    id: T,
    index: number,
    onEdit?: (id: T) => void,
    onRemove?: (id: T) => void,
    sortable: boolean,
};

export default class Item<T> extends React.PureComponent<Props<T>> {
    static defaultProps = {
        disabled: false,
        sortable: true,
    };

    createDragHandle() {
        const {sortable} = this.props;

        const handle = ({className, children}: Object) => (
            <span className={className}>{children}</span>
        );

        if (!sortable) {
            return handle;
        }

        return SortableHandle(handle);
    }

    handleEdit = () => {
        const {id, onEdit} = this.props;

        if (onEdit) {
            onEdit(id);
        }
    };

    handleRemove = () => {
        const {id, onRemove} = this.props;

        if (onRemove) {
            onRemove(id);
        }
    };

    render() {
        const {
            children,
            disabled,
            index,
            onEdit,
            onRemove,
            sortable,
        } = this.props;

        const DragHandle = this.createDragHandle();

        const itemClass = classNames(
            itemStyles.item,
            {
                [itemStyles.disabled]: disabled,
            }
        );

        const dragHandleClass = classNames(
            itemStyles.dragHandle,
            {
                [itemStyles.sortable]: sortable,
            }
        );

        return (
            <div className={itemClass}>
                <DragHandle className={dragHandleClass}>
                    {sortable && <Icon name={DRAG_ICON} />}
                    <span className={itemStyles.index}>{index}</span>
                </DragHandle>
                <div className={itemStyles.content}>
                    {children}
                </div>
                <div className={itemStyles.buttons}>
                    {onEdit &&
                        <button className={itemStyles.button} onClick={this.handleEdit} type="button">
                            <Icon name="su-pen" />
                        </button>
                    }
                    {onRemove &&
                        <button className={itemStyles.button} onClick={this.handleRemove} type="button">
                            <Icon name="su-trash-alt" />
                        </button>
                    }
                </div>
            </div>
        );
    }
}
