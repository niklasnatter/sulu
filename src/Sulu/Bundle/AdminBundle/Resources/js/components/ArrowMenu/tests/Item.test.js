// @flow
import React from 'react';
import {render, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Item from '../Item';

test('Render default Item', () => {
    const {container} = render(<Item value="test">Test Item</Item>);

    expect(container).toMatchSnapshot();
});

test('Render active Item', () => {
    const {container} = render(<Item active={true} icon="fa-home" value="house">My House</Item>);

    expect(container).toMatchSnapshot();
});

test('Render disabled Item', () => {
    const {container} = render(<Item active={true} disabled={true} icon="fa-home" value="house">My Item</Item>);

    expect(container).toMatchSnapshot();
});

test('Clicking the left and right button inside the header should call the right handler', async () => {
    const clickHandler = jest.fn();
    render(<Item active={true} icon="fa-home" onClick={clickHandler} value="house">My House</Item>);

    const user = userEvent.setup();
    await user.click(screen.getByText('My House'));

    expect(clickHandler).toBeCalledWith('house');
});

test('Clicking the disabled Item should not call a handler', async () => {
    const clickHandler = jest.fn();
    render(<Item active={false} disabled={true} icon="fa-home" onClick={clickHandler} value="house">My House</Item>);

    const user = userEvent.setup();
    await user.click(screen.getByText('My House'));

    expect(clickHandler).not.toBeCalled();
});
