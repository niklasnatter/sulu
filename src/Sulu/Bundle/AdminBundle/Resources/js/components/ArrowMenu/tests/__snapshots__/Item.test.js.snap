// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render active Item 1`] = `
<button
  class="item active"
  type="button"
>
  <span
    class="icon"
  >
    <span
      aria-label="fa-home"
      class="fa fa-home icon"
    />
  </span>
  <span>
    My House
  </span>
</button>
`;

exports[`Render default Item 1`] = `
<button
  class="item"
  type="button"
>
  <span
    class="icon"
  />
  <span>
    Test Item
  </span>
</button>
`;

exports[`Render disabled Item 1`] = `
<button
  class="item active"
  disabled=""
  type="button"
>
  <span
    class="icon"
  >
    <span
      aria-label="fa-home"
      class="fa fa-home icon"
    />
  </span>
  <span>
    My Item
  </span>
</button>
`;
